File: 2024-08/16.py
# Problem: 624. Maximum Distance in Arrays
# Link: https://leetcode.com/problems/maximum-distance-in-arrays/
# Beats: 39.94% (Runtime: 1358 ms), 51.44% (Memory: 42.52 MB)
# Time Complexity: O(m) - The algorithm processes each of the m arrays once.
# Space Complexity: O(1) - The space used is constant.

from typing import List

class Solution:
    def maxDistance(self, arrays: List[List[int]]) -> int:
        # Initialize minimum and maximum values from the first array
        min_val = arrays[0][0]
        max_val = arrays[0][-1]
        max_distance = 0
        
        # Iterate over the remaining arrays
        for array in arrays[1:]:
            # Calculate the possible maximum distances using the current array
            max_distance = max(max_distance, abs(array[-1] - min_val), abs(max_val - array[0]))
            # Update min_val and max_val to include the current array
            min_val = min(min_val, array[0])
            max_val = max(max_val, array[-1])
        
        return max_distance

# Example usage:
solution = Solution()

# Test case 1
arrays1 = [[1, 2, 3], [4, 5], [1, 2, 3]]
print(solution.maxDistance(arrays1))  # Output: 4

# Test case 2
arrays2 = [[1], [1]]
print(solution.maxDistance(arrays2))  # Output: 0

File: 2024-09/16.cpp
// Problem: 539. Minimum Time Difference
// Link: https://leetcode.com/problems/minimum-time-difference/
// Beats: 53.24% with 13 ms runtime, 30.60% with 17.92 MB memory usage

/*
Time Complexity: O(n log n)
- The time complexity is dominated by the sorting of the `timePoints` array, which takes O(n log n). After sorting, calculating the differences between adjacent times takes O(n), making the overall complexity O(n log n).

Space Complexity: O(1)
- The space complexity is O(1), ignoring the space required to store the input, as we perform operations in-place on the sorted list and use only a few variables for calculations.
*/

#include <iostream>
#include <vector>
#include <string>
#include <climits>
#include <algorithm> // for sort

using namespace std;

class Solution
{
public:
    // Function to calculate the time difference between two time points
    int timeDifference(string t1, string t2)
    {
        int minutes1 = stoi(t1.substr(0, 2)) * 60 + stoi(t1.substr(3, 2)); // Convert t1 to minutes
        int minutes2 = stoi(t2.substr(0, 2)) * 60 + stoi(t2.substr(3, 2)); // Convert t2 to minutes

        // Return the minimum of the direct difference and the wrap-around difference (24*60 = 1440 minutes in a day)
        return min(abs(minutes2 - minutes1), 1440 - abs(minutes2 - minutes1));
    }

    int findMinDifference(vector<string> &timePoints)
    {
        sort(timePoints.begin(), timePoints.end()); // Sort the time points
        int minDifference = INT_MAX;                // Initialize the minimum difference with a large value
        int n = timePoints.size();

        // Calculate the minimum time difference between adjacent time points
        for (int i = 0; i < n; i++)
        {
            minDifference = min(minDifference, timeDifference(timePoints[i], timePoints[(i + 1) % n]));
            if (minDifference == 0) // If minimum difference is 0, return immediately as we cannot get a smaller value
                return minDifference;
        }

        return minDifference;
    }
};

int main()
{
    Solution solution;

    // Example 1
    vector<string> timePoints1 = {"23:59", "00:00"};
    cout << "Example 1 Output: " << solution.findMinDifference(timePoints1) << endl; // Expected output: 1

    // Example 2
    vector<string> timePoints2 = {"00:00", "23:59", "00:00"};
    cout << "Example 2 Output: " << solution.findMinDifference(timePoints2) << endl; // Expected output: 0

    return 0;
}
